// @redocly/config/lib/feedback-config-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.feedbackConfigSchema = exports.reasonsSettingsSchema = void 0;
const reasonsBaseProperties = {
    hide: {
        type: 'boolean',
        default: false,
    },
    component: {
        type: 'string',
        enum: ['radio', 'checkbox'],
        default: 'checkbox',
    },
    label: { type: 'string' },
    items: { type: 'array', items: { type: 'string' } },
};
exports.reasonsSettingsSchema = {
    type: 'object',
    properties: reasonsBaseProperties,
    additionalProperties: false,
};
exports.feedbackConfigSchema = {
    type: 'object',
    properties: {
        hide: {
            type: 'boolean',
            default: false,
        },
        type: {
            type: 'string',
            enum: ['rating', 'sentiment', 'comment', 'reasons', 'mood', 'scale'],
            default: 'sentiment',
        },
        settings: {
            type: 'object',
            properties: {
                label: { type: 'string' },
                submitText: { type: 'string' },
                buttonText: { type: 'string' },
                component: {
                    type: 'string',
                    enum: ['radio', 'checkbox'],
                    default: 'checkbox',
                },
                items: { type: 'array', items: { type: 'string' }, minItems: 1 },
                leftScaleLabel: { type: 'string' },
                rightScaleLabel: { type: 'string' },
                reasons: {
                    type: 'object',
                    properties: Object.assign(Object.assign({}, reasonsBaseProperties), { like: exports.reasonsSettingsSchema, dislike: exports.reasonsSettingsSchema, satisfied: exports.reasonsSettingsSchema, neutral: exports.reasonsSettingsSchema, dissatisfied: exports.reasonsSettingsSchema }),
                    additionalProperties: false,
                },
                comment: {
                    type: 'object',
                    properties: {
                        hide: {
                            type: 'boolean',
                            default: false,
                        },
                        label: { type: 'string' },
                        likeLabel: { type: 'string' },
                        dislikeLabel: { type: 'string' },
                        satisfiedLabel: { type: 'string' },
                        neutralLabel: { type: 'string' },
                        dissatisfiedLabel: { type: 'string' },
                    },
                    additionalProperties: false,
                },
            },
            additionalProperties: false,
        },
    },
    additionalProperties: false,
};
